### PR TITLE
Adding sample grouping to identify forward key frames

### PR DESCRIPTION
**Pull request by cconcolato**

### CODE REVIEW

File: index.bs [hunk 1]
```diff
@@ -203,7 +201,7 @@ Unlike many video standards, AV1 does not distinguish the display order from the
 
 When a temporal unit contains more than one frame, the sample corresponding to that temporal unit should be marked using the [=AV1MultiFrameSampleGroupEntry=].
 
-The grouping_type_parameter is not defined for the SampleToGroupBox with grouping types defined in this specification (i.e. 'av1m', 'av1s', 'av1M').
+The grouping_type_parameter is not defined for the SampleToGroupBox with grouping types defined in this specification (i.e. 'av1m', 'av1s', 'av1M', 'av1f').
```
Comment (tdaede) [inline index.bs]: Unrelated change?

File: index.bs [hunk 2]
```diff
@@ -229,6 +227,39 @@ The <dfn>AV1MultiFrameSampleGroupEntry</dfn> documents samples that contain mult
...
+```
+class AV1ForwardAccessSampleGroupEntry extends VisualSampleGroupEntry('av1f') {
+  unsigned int(1) initial_display;
+  unsigned int(7) fwd_distance;
+}
+```
+
+### Semantics ### {#forwardaccesssamplegroupentry-semantics}
+
+The <dfn>initial_display</dfn> fields indicates if the processing of this sample will produce an output frame or not.
```
Comment (tdaede) [inline index.bs]: This is for a sample group, not sample. Did you intend to mean whether *any* sample in the group will produce an output frame?

File: index.bs [hunk 3]
```diff
@@ -145,8 +145,8 @@ class AV1CodecConfigurationBox extends FullBox('av1C', version = 0, 0){
 }
 
 aligned (8) class AV1CodecConfigurationRecord {
-  unsigned int (6) reserved = 0;
-  unsigned int (2) OBU_length_mode;
+  unsigned int (7) reserved = 0;
```
Comment (tomfinegan) [inline index.bs]: Seems this stacked w/the OBU_LENGTH patch. Needs the 4 here.

File: index.bs (hunk 2 above)
Comment (tomfinegan) [reply index.bs]: +1. Shouldn't it be something like:

indicates if the processing of the first sample in the group will produce an output frame or not

File: index.bs (hunk 1 above)
Comment (cconcolato) [reply index.bs]: No it is. The 4CC for the new sample group is av1f. Maybe I should simply remove the text between parenthesis. That would avoid updating it every time a new Sample Group is defined ...

File: index.bs [hunk 4]
```diff
@@ -237,6 +227,39 @@ The <dfn>AV1MultiFrameSampleGroupEntry</dfn> documents samples that contain mult
...
+  unsigned int(1) initial_display;
+  unsigned int(7) fwd_distance;
+}
+```
+
+### Semantics ### {#forwardaccesssamplegroupentry-semantics}
+
+The <dfn>initial_display</dfn> fields indicates if the processing of this sample will produce an output frame or not.
+
+The <dfn>fwd_distance</dfn> field indicates the number of samples whose decoding can be skipped before encountering the sample outputting the next frame. 0 means the next sample.
```
Comment (tomfinegan) [inline index.bs]: Will muxers have to look at frame headers to determine this value?

File: index.bs (hunk 1 above)
Comment (tdaede) [reply index.bs]: Oh. Somehow I misparsed it as related to AVIF-the-image-container.

File: index.bs (hunk 2 above)
Comment (cconcolato) [reply index.bs]: A sample group is the mechanism to attach the same property to a set of samples. In this case, each sample marked in the sample group produces an output frame (or not) and is followed N frames after by a show existing frame. Can you suggest better text?

Comment (cconcolato) [reply index.bs]: @tomfinegan The concept of sample group is not used to describe a set of consecutive samples. There is no notion of first. All samples in a sample group have the same characteristics.

File: index.bs (hunk 3 above)
Comment (cconcolato) [reply index.bs]: Yes, I must have messed up when I created the branch ... I will make sure the OBU_LENGTH commit is removed from this branch when I merge.

File: index.bs (hunk 4 above)
Comment (cconcolato) [reply index.bs]: Indeed. Once a muxer parses a Key Frame with show_frame=0, it would have to look forward in the stream to see if there is a show_existing_frame pointing to it, to be able to set this distance. The idea is that the demuxer would not have to do this job.

File: index.bs (hunk 2 above)
Comment (tomfinegan) [reply index.bs]: OK, this makes more sense knowing that. Thanks.

Comment (tdaede) [reply index.bs]: OK. I was a bit confused because I thought it was normally mandatory for all non-fwd-kf sample groups to output a frame, so it didn't seem important to include this bit if it's already labeled av1f. Can't you assume it's 0 if your sample is av1f?

Comment (cconcolato) [reply index.bs]: Because the KF has show_frame = 0 it has to be bundled in the TU/Sample with another frame that actually is showable. That frame may or may not be decodable with the preceding key frame only. That bit was for that. There are also complex configurations using INTRA ONLY FRAMEs that could be used to achieve the same effect with show_frame = 0 or 1. That's why the sample group is not called ForwardKeyFrame as I think it can be used also in different configurations.

File: index.bs (hunk 4 above)
Comment (cconcolato) [reply index.bs]: Generating this sample group is indeed complex, but it is optional so that is less problematic. To generate it, I think either you have a good integration with your encoder and you will be able to generate it easily or you would have to do complex parsing and reference buffer management. Maybe the frame ids can help, but you can also do without by tracking buffer references.

File: index.bs (hunk 2 above)
Comment (tdaede) [reply index.bs]: OK. I think it now makes sense why the bit is there, but not happy with the name or description. I will try to come up with something better, but it doesn't have to block this pr.

**Comment by cconcolato** [comment]
This PR is pending the discussions on the Random Access Point, I won't merge it for now.
